File: Python/big-o-rule-1/main.py
from profile import run as performance_run

NEMO = ['nemo']

EVERYONE = [
    'dory', 'bruce', 'marlin', 'gill', 'bloat', 'nigel', 'squirt',
    'darla', 'hank', 'nemo'
]  # Worst Case Scenario BIG O Rule 1.

LARGE = ['nemo'] * 100000


def find_nemo(array):
    """
    Searches for the string 'nemo' in the given array.
    Time complexity: O(n) -> Linear Time

    Parameters:
    array (list): The array to search in.

    Returns:
    None
    """
    for i in range(len(array)):
        if array[i] == 'nemo':
            print('Found NEMO!')
            # break
            # Even if you find nemo before the end of the array,
            # the time complexity is still O(n). (Worst Case)


performance_run('find_nemo(LARGE)')
# find_nemo(LARGE)

File: src/big-o-rule-1/main.go
package main

import "fmt"

/*
Big O Rule 1: Worst Case

The worst-case scenario is the scenario in which the algorithm takes the longest time to execute.
It is important to consider the worst-case scenario when analyzing the time complexity of an algorithm.

findNemo searches for the string "a" in the given slice of strings.
If "a" is found, it prints "Found Nemo!" to the console.

Time complexity: O(n)
Space complexity: O(1)

Parameters:

	nemo: a slice of strings

Return:

	nil

Example:

	findNemo([]string{"nemo", "dory", "bruce", "marlin", "gill", "bloat", "nigel", "squirt", "darla", "hank"})
*/
func findNemo(nemo []string) {
	for i := 0; i < len(nemo); i++ {
		if nemo[i] == "nemo" {
			fmt.Println("Found Nemo!")
		}
	}
}

func main() {
	nemo := []string{
		"nemo", "dory", "bruce", "marlin", "gill", "bloat", "nigel", "squirt", "darla", "hank",
	}
	findNemo(nemo)
	large := make([]string, 1000000)
	for i := 0; i < len(large); i++ {
		large[i] = "nemo"
	}
	findNemo(large)
}
